### PR TITLE
Explicitly type GraphQL Schema using Gatsby Schema Customisation API

### PR DESCRIPTION
**Pull request by WillTheVideoMan**
At present, `gatsby-source-ghost` uses fake nodes and dummy content to fill in the schema. Using the Gatsby schema customisation API, the schema can be explicitly typed and fully reflect the Ghost API V3 spec. 

Using the dummy nodes method, when changes are made in the Ghost Admin panel to cause the API to return null for a given field, in some cases the schema simply drops the field rather than also returning null (#42). This adds friction to development, builds and is generally confusing to developers who expect the schema to always reflect the Ghost V3 API spec. 

Introducing an explicitly typed schema means developers can make queries without worrying about handling a field that doesn't exist. The action of users in the Ghost admin panel should not halt builds altogether. It would also open up possibilities for schema rebuilding (#44), and might help in getting Ghost to become a first-class integration (#15) with Gatsby. 

This PR eliminates fake nodes and dummy content, and instead uses an explicitly typed schema and the Gatsby Schema Customisation API to generate a reliable set of Gatsby nodes. 

I would love to collaborate towards making this change perfect. If there are any issues in the changes I have implemented, let's work them out and see what we can do! My fork seems to work well for me, and is passing, but then again something always slips under the net. For example, if some of the field types are incorrect, or I have missed the point of some functions in my comments, let me know.

Cheers, and I hope you are all well. 

P.S. In the mean time, for anyone struggling with these issues, I have made a scoped package over on NPM. Just use `npm i @willthevideoman/gatsby-source-ghost` and update the `gatsby-config.js` for a drop-in replacement.

### CODE REVIEW

**Comment by WillTheVideoMan** [comment]
I have stumbled across another source plugin which seems to also cover most of these bases and includes caching and logging too. Looks like a super example of what could be done with this source plugin!

https://www.npmjs.com/package/jamify-source-ghost

Cheers!

File: ghost-schema.js
Comment (pieh) [inline ghost-schema.js:48]:
```suggestion
    email_subject: String
    mobiledoc: String
    plaintext: String
    page: Boolean
```

This relates to comments about fields from "fake" post/page missing in GraphQL types - I don't know if this is correct - both of all those fields are actually there in GhostAPI response and if those should be nullable or not (so I didn't add `!` in types)

File: ghost-schema.js
Comment (pieh) [inline ghost-schema.js:86]:
```suggestion
    email_subject: String
    mobiledoc: String
    plaintext: String
    page: Boolean
```

This relates to comments about fields from "fake" post/page missing in GraphQL types - I don't know if this is correct - both of all those fields are actually there in GhostAPI response and if those should be nullable or not (so I didn't add `!` in types)

File: gatsby-node.js
Comment (pieh) [inline gatsby-node.js:114]: This will result in possible breaking change - previously this field would result in empty string (if `_.isNil` check is truthy), with removal this will result in `null`. This can cause problems if someone tries to some string manipulation assuming that this field is never null

File: gatsby-node.js
Comment (pieh) [inline gatsby-node.js:46]: This will result in possible breaking change - previously this field would result in empty string (if _.isNil check is truthy), with removal this will result in null. This can cause problems if someone tries to some string manipulation assuming that this field is never null

**Review by pieh — CHANGES_REQUESTED** [review]
I'm not maintainer of the plugin, but I did leave couple of comments and suggestions - some of them can be just applied, other have some questions that need more context on Ghost API to decide what to do with them.

Lastly - using schema customization might be considered breaking change, because it requires certain versions of `gatsby` (particularly `@link` directive was added in `gatsby@2.5.0`), so depending which gatsby versions plugin is meant to support - it might require major version bump (or not if it already requires newer `gatsby`)

**Review by WillTheVideoMan — COMMENTED** [review]
Thanks for the review, there were certainly lots of things I had missed. There are some queries I have in regard to certain fields no-longer existing in the Ghost Content API v3, and how we deal with preventing breaking changes in these areas. 

I look forward to hearing your thoughts! Cheers again.

File: ghost-schema.js
Comment (WillTheVideoMan) [reply ghost-schema.js:48]: On review, it looks like `page` is only set in the ghost pages API endpoint shape, as it is null in the ghost post endpoint shape. This means `page` should also be nullable.

File: gatsby-node.js
Comment (WillTheVideoMan) [reply gatsby-node.js:46]: On review, I remember why I removed this check. If there is no code injection in either the head or foot, the null check above on lines 78-80 will return null rather than an empty string. This means it is possible for `codeinjection_styles` to be null, that clients can already expect to deal with it, and so might not be a breaking change. Let me know what you think.

File: gatsby-node.js
Comment (WillTheVideoMan) [reply gatsby-node.js:114]: In reference to above, this does not match the return pattern of `transformCodeinjection()`. This would be a breaking change. Should this be normalised to be entirely nullable or entirely non-nullable?

**Comment by aileen** [comment]
Hey @WillTheVideoMan 👋 That's a great improvement! Happy to merge when the conflicts are resolved and the missing types are added. Also cheers to @pieh for helping out here 😊

**Comment by WillTheVideoMan** [comment]
Hi @aileen! 

The above changes and conflicts have hopefully been resolved. Pending your review, I hope things are looking for a merge! Please do let me know if there is anything else I need to do or tweak, and I will hop on that straight away.

Cheers!

**Review by aileen — APPROVED** [review]
This works great! Thanks for the work 🙏

**Pull request merged**
